src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -26,80 +26,198 @@
+
+  private final ScheduledExecutorService recycler = Executors.newScheduledThreadPool(1);
+
+  private static class Epoch<T> {
+    private final T inner;
+    private final int epoch;
+
+    public Epoch(T inner, int epoch) {
+      this.inner = inner;
+      this.epoch = epoch;
+    }
+
+    public T get() {
+      return inner;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other == this) {
+        return true;
+      } else if (other instanceof Epoch) {
+        Epoch<?> o = (Epoch<?>) other;
+        return o.epoch == epoch;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      return epoch;
+    }
+  }
+
+  @VisibleForTesting
+  public static class CertWatcher implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(CertWatcher.class);
+    private final String path;
+    private final ArrayList<String> targets = new ArrayList<>();
+    private final BlockingDeque<Boolean> notify = new LinkedBlockingDeque<>();
+    private final ExecutorService executorService = Executors.newFixedThreadPool(2);
+
+    public CertWatcher(String path) {
+      if (path == null) {
+        throw new IllegalArgumentException("the base path of certs is missing");
+      }
+      this.path = path;
+      notify.push(true);
+      start();
+    }
+
+    private void watchLoop() {
+      try {
+        WatchService watchService = FileSystems.getDefault().newWatchService();
+        Path p = Paths.get(path);
+        p.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+
+        while (true) {
+          WatchKey key = watchService.take();
+          boolean changed = false;
+          OUTER:
+          for (WatchEvent<?> event : key.pollEvents()) {
+            for (String target : targets) {
+              logger.info("detected file change: {}", event.context());
+              if (event.context().toString().equals(target)) {
+                changed = true;
+                break OUTER;
+              }
+            }
+          }
+          if (changed) {
+            notify.offer(true);
+          }
+          if (!key.reset()) {
+            String msg = "fail to reset watch key";
+            logger.warn(msg);
+            throw new RuntimeException(msg);
+          }
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private void start() {
+      executorService.submit(this::watchLoop);
+    }
 
-    protected abstract SslContextBuilder createSslContextBuilder();
+    public CertWatcher addTarget(String target) {
+      targets.add(target);
+      return this;
+    }
 
-    public SslContextBuilder reload() {
-      if (isModified()) {
-        logger.info("reload ssl context");
-        return createSslContextBuilder();
+    public boolean isModified() throws Exception {
+      boolean modified = !notify.isEmpty();
+      if (modified) {
+        notify.take();
```

Comment:
Maybe better to use `clear`? If there are multiple `true`s in it, no need to trigger reload multiple times.